### ISSUE TITLE
Specialization doesn't trigger for these cases

### ISSUE THREAD

**Issue by spinda** [rust-lang/rust#38642]
As discussed in https://github.com/rust-lang/rfcs/issues/1825, specialization should allow the following to compile:

(original impetus, using the `!` type)

```rust
#![feature(specialization)]
#![feature(never_type)]

trait Unto<T> {
    fn unto(self) -> T;
}

impl<T> Unto<T> for T {
    default fn unto(self) -> T {
        self
    }
}

impl<T> Unto<T> for ! {
    fn unto(self) -> T {
        self
    }
}

fn main() {}
```

(general case, without `!`)

```rust
#![feature(specialization)]

trait Unto<T> {
    fn unto(self) -> T;
}

impl<T> Unto<T> for T {
    default fn unto(self) -> T {
        self
    }
}

impl<T> Unto<T> for u32 {
    fn unto(self) -> T {
        panic!("aaah")
    }
}

fn main() {}
```

cc @canndrew @aturon

**Comment by spinda** [comment]
I suppose I should tag #31844 here as well.

**Comment by canndrew** [comment]
Thanks for opening this. Actually I don't think those examples should compile as described because neither impl is contained by the other. It should work if you explicitly write an impl for the overlap though (but currently it doesn't)

```rust
#![feature(specialization)]
#![feature(never_type)]

trait Unto<T> {
    fn unto(self) -> T;
}

impl<T> Unto<T> for T {
    default fn unto(self) -> T {
        self
    }
}

impl<T> Unto<T> for ! {
    default fn unto(self) -> T {
        self
    }
}

impl Unto<!> for ! {
    fn unto(self) -> ! {
        self
    }
}
```

**Comment by spinda** [comment]
After looking over the `rustc` source, this is what happens now when compiling that example:

1. Attempt to insert `impl<T> Unto<T> for T` into the specialization graph. No overlaps detected, so insertion succeeds.

2. Attempt to insert `impl<T> Unto<T> for !` into the specialization graph. An overlap is detected with `impl<T> Unto<T> for T` currently in the specialization graph; specifically, `!: Unto<!>` is the overlapping case. Neither `impl` contains the other, so `impl<T> Unto<T> for !` is not inserted into the graph and an error is emitted.

3. Attempt to insert `impl Unto<!> for !` into the specialization graph. An overlap is detected with `impl<T> Unto<T> for T` currently in the graph. `impl Unto<!> for !` is contained by `impl<T> Unto<T> for T` so specialization kicks in and insertion is permitted.

4. Down the line, compilation is aborted due to the error emitted in step 2.

This is what I think would need to happen:

1. Attempt to insert `impl<T> Unto<T> for T` into the specialization graph. No overlaps detected, so insertion succeeds.

2. Attempt to insert `impl<T> Unto<T> for !` into the specialization graph. An overlap is detected with `impl<T> Unto<T> for T` currently in the specialization graph; specifically, `!: Unto<!>` is the overlapping case. Because all associated items in both `impl`s are marked with `default`, `rustc` notes that a specialization could resolve this overlap. `!: Unto<!>` is added to an internal conflict list in `OverlapChecker` and `impl<T> Unto<T> for !` is tentatively inserted into the specialization graph.

3. Attempt to insert `impl Unto<!> for !` into the specialization graph. An overlap is detected with `impl<T> Unto<T> for T` currently in the graph. `impl Unto<!> for !` is contained by `impl<T> Unto<T> for T` so specialization kicks in and insertion is permitted. This eliminates `!: Unto<!>` from the conflict list.

4. If anything is left in the conflict list at the end of overlap checking, errors are now emitted and compilation halts.

Step 2 seems a bit fiddly. cc @aturon

**Comment by arielb1** [comment]
Lattice specialization (aka anything which requires intersection impls) is not yet implemented.

**Comment by canndrew** [comment]
@arielb1 Is there an issue open for lattice specialization somewhere?

**Comment by arielb1** [comment]
I don't think there is even an RFC.
